cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1195,6 +1195,43 @@ struct ResolveIfElseExec<NullType, AllocateMem> {
   }
 };
 
+template <typename ResolverForOtherTypes>
+Result<TypeHolder> ResolveDecimalCaseType(KernelContext* ctx,
+                                          const std::vector<TypeHolder>& types,
+                                          ResolverForOtherTypes&& resolver) {
+  if (!HasDecimal(types)) {
+    return resolver(ctx, types);
+  }
+
+  int32_t max_precision = 0, max_scale = 0;
+  for (auto& type : types) {
+    if (is_floating(type.id()) || is_integer(type.id())) {
+      return Status::Invalid("Need to cast numeric types containing decimal types");
+    } else if (is_decimal(type.id())) {
+      const auto& decimal_type = checked_cast<const arrow::DecimalType&>(*type);
+      if (decimal_type.precision() < max_precision || decimal_type.scale() < max_scale) {
```

Comment:
< 0?

Reply:
The returned invalid status used in the logic :
https://github.com/bkietz/arrow/blob/c25866bc59e30e43e8dc3b05f3973d48074c3594/cpp/src/arrow/compute/expression.cc#L551-L552

Let the expression call go into `DispatchBest` so they can be cast correctly.

Reply:
In `BindNonRecursive`:
`FinishBind` will do output type resolve. For the first time go into the resolver, our decimal related types haven't cast, we could check if the decimal types have different precision and scale. 

And then return invalid to make the `BindNonRecursive` go into `IfElse`'s related function's `DispatchBest` and do cast.

After that, when we go into type resolver in `BindNoRecursive` second time, we could find the decimal types already cast to same precision and scale, so we could find the correct result output type.